### PR TITLE
Added .mcap format to recording

### PR DESCRIPTION
**Pull request by ZigiMigi**
Added class Mcap to depthai_record and  added option to record files into .mcap format

### CODE REVIEW

File: gen2-record-replay/libraries/depthai_record.py [hunk 1]
```diff
@@ -31,43 +39,63 @@ def __init__(self, path: Path, device) -> None:
...
+
+                        # frames[name].tofile(files[name])
+
+                    else:
+                        if name not in mcaps:  # File wasn't created yet
+                            create_mcap_file(name)
 
-                    # if self.rotate != -1: # Doesn't work atm
-                    #     frames[name] = cv2.rotate(frames[name], self.rotate)
+                        mcaps[name].imageSave(frames[name])
```
Comment (Erol444) [inline gen2-record-replay/libraries/depthai_record.py:84]: Can't all streams be saved in a single MCAP file? If that is possible, I assume schemas will get a bit more complicated (as you would need to dynamically create it), but we should go for that.

File: gen2-record-replay/libraries/depthai_record.py [hunk 2]
```diff
@@ -31,43 +39,63 @@ def __init__(self, path: Path, device) -> None:
         calibData.eepromToJsonFile(str(self.path / "calib.json"))
 
         self.convert_mp4 = False
+        self.mcap = mcap
 
     def run(self):
         files = {}
+        mcaps = {}
```
Comment (Erol444) [inline gen2-record-replay/libraries/depthai_record.py:46]: If we have only one .mcap file, this dictionary isn't needed.

File: gen2-record-replay/libraries/depthai_record.py [hunk 3]
```diff
@@ -31,43 +39,63 @@ def __init__(self, path: Path, device) -> None:
...
                 ext = 'h265' if self.quality == EncodingQuality.LOW else 'mjpeg'
                 files[name] = open(str(self.path / f"{name}.{ext}"), 'wb')
             # if name == "color": fourcc = "I420"
             # elif name == "depth": fourcc = "Y16 " # 16-bit uncompressed greyscale image
             # else : fourcc = "GREY" #Simple, single Y plane for monochrome images.
-            # files[name] = VideoWriter(str(path / f"{name}.avi"), VideoWriter_fourcc(*fourcc), fps, sizes[name], isColor=name=="color")
+            # files[name] = VideoWriter(str(path / f"{name}.avi"), VideoWriter_fourcc(*fourcc), fps, sizes[name],
+            # isColor=name=="color")
+
+        def create_mcap_file(name):
```
Comment (Erol444) [inline gen2-record-replay/libraries/depthai_record.py:60]: Dynamically create MCAP schema and create only one .mcap file.

File: gen2-record-replay/libraries/depthai_record.py [hunk 4]
```diff
@@ -242,3 +270,193 @@ def stream_out(name, fps, out, noEnc=False):
         self.pipeline = pipeline
         return pipeline, nodes
 
+
+class Mcap:
```
Comment (Erol444) [inline gen2-record-replay/libraries/depthai_record.py:274]: Move the class to libraries/depthai_mcap.py, and rename it to `DepthAiMcap` (similar to the `DepthAiBags`)

File: gen2-record-replay/libraries/depthai_record.py [hunk 5]
```diff
@@ -242,3 +270,193 @@ def stream_out(name, fps, out, noEnc=False):
...
+        for point in points:
+            buf += struct.pack('f', float(point[0]))
+            buf += struct.pack('f', float(point[1]))
+            buf += struct.pack('f', float(point[2]))
+
+        data = base64.b64encode(buf).decode("ascii")
+
+        tmpTime = time.time_ns()
+        sec = math.trunc(tmpTime / 1e9)
+        nsec = tmpTime - sec
```
Comment (Erol444) [inline gen2-record-replay/libraries/depthai_record.py:424]: Are these variables even used?

File: gen2-record-replay/libraries/depthai_record.py [hunk 6]
```diff
@@ -242,3 +270,193 @@ def stream_out(name, fps, out, noEnc=False):
...
+        # read from .jpeg format to buffer of bytes
+        # byte_im = im_buf_arr.tobytes()
+        byte_im = img.tobytes()
+
+        # data must be encoded in base64
+        data = base64.b64encode(byte_im).decode("ascii")
+
+        tmpTime = time.time_ns()
+        sec = math.trunc(tmpTime / 1e9)
+        nsec = tmpTime - sec
```
Comment (Erol444) [inline gen2-record-replay/libraries/depthai_record.py:391]: Remove.

File: gen2-record-replay/libraries/depthai_record.py [hunk 7]
```diff
@@ -242,3 +270,193 @@ def stream_out(name, fps, out, noEnc=False):
...
+
+        self.writer.add_message(
+            channel_id=self.channel_id,
+            log_time=tmpTime,
```
Comment (Erol444) [inline gen2-record-replay/libraries/depthai_record.py:395]:
```suggestion
            log_time=time.time_ns(),
```

**Review by Erol444 — CHANGES_REQUESTED** [review]

File: gen2-record-replay/libraries/depthai_record.py (hunk 5 above)
Comment (ZigiMigi) [reply gen2-record-replay/libraries/depthai_record.py:424]: yes they are, they are sent in the header of the message

File: gen2-record-replay/libraries/depthai_record.py (hunk 6 above)
Comment (ZigiMigi) [reply gen2-record-replay/libraries/depthai_record.py:391]: Variables are needed.

File: gen2-record-replay/libraries/depthai_record.py (hunk 7 above)
Comment (ZigiMigi) [reply gen2-record-replay/libraries/depthai_record.py]: This is needed, as the time must be same as time in header (calling it 2 times produces 2 different times.

**Comment by Erol444** [comment]
TODO: when recording depth and using mcap, use pointCloudSave()

**Comment by Erol444** [comment]
I will be adding depth->pcl conversion and add these commits to the record refactor PR: https://github.com/luxonis/depthai-experiments/pull/369

**Pull request closed without merging**